### PR TITLE
[oneDNN] Extend shape of elementwise_add/sub for swin_transformer

### PR DESCRIPTION
**Pull request by LLee233**
### PR types

Performance optimization

### PR changes

Others

### Description

This PR together with https://github.com/PaddlePaddle/Paddle/pull/59394 are aiming for optimizing **swin_transformer** in https://github.com/PaddlePaddle/Paddle/issues/59252.

The reported regression is caused by perf regression of elementwise_add/sub and fused matmul. Due to rare input shapes, these two kernels can not utilize corresponding broadcast on oneDNN, hence the regression. Here we workaround to extend shape of input/residual_data and keep utilizing optimization of oneDNN

### CODE REVIEW

**Comment by LLee233** [comment]
Hi [@yuanlehome](https://github.com/yuanlehome), could you please help check CI coverage? Since I saw the result is kind of unexpected, where header file got 0% coverage... However the upper level code (in elementwise_kernel.cc) got called thousands of times, and there is no condition causing return or something before entering header file. Thanks!

**Comment by LLee233** [comment]
@xinyu-intel, @vivienfanghuagood, @yuanlehome, hi, would you mind helping review this PR? Thanks~

File: paddle/phi/kernels/onednn/elementwise_kernel.cc [hunk 1]
```diff
@@ -77,8 +77,18 @@ void ElementwiseKernel(const OneDNNContext& dev_ctx,
     std::swap(non_const_x, non_const_y);
   }
 
-  const auto src_x_memory = handler.AcquireSrcMemory(non_const_x);
-  const auto src_y_memory = handler.AcquireSecondSrcMemory(non_const_y);
+  const auto src_x_memory =
+      handler.swin_case ? (x.numel() == y.numel()
```
Comment (yuanlehome) [inline paddle/phi/kernels/onednn/elementwise_kernel.cc:81]: 这里的x.numel() == y.numel()判断和
image
是不是重复了？

Comment (LLee233) [reply paddle/phi/kernels/onednn/elementwise_kernel.cc:81]: Hi @yuanlehome, 这里虽然也对x和y的numel进行了判定，但原因不太一样。在swin tranformer里收到影响的elementwise op有两个，一个是add一个是sub，并且两个op受到影响的原因不同，所以要根据x和y的numel数量作进一步判断。在下面的code里（也就是在handler里），这个判断是用于决定当前受影响的是add还是sub，根据结果对不同的src md进行处理。但这个比较结果无法直接传递在上面的code里（kernel里），或者说想要传递可能需要增加额外的变量，感觉必要性不大。因此，我在上面的code再次进行了比较，再根据结果给`AcquireExtendSrcMemory` 传入不同的参数。不过如果觉得numel开销比较大的话，我可以试着用别的条件替代。

Comment (yuanlehome) [reply paddle/phi/kernels/onednn/elementwise_kernel.cc:81]: 好的了解，numel开销应该比较小的

**Review by yuanlehome — APPROVED** [review]

File: paddle/phi/kernels/onednn/elementwise_kernel.cc (hunk 1 above)
Comment (LLee233) [reply paddle/phi/kernels/onednn/elementwise_kernel.cc:81]: 嗯嗯好的~

**Review by xinyu-intel — APPROVED** [review]

**Pull request merged**